src/lib.rs

```diff
@@ -89,27 +92,24 @@ impl<T> ReusableBoxFuture<T> {
     where
         F: Future<Output = T> + Send + 'static,
     {
-        // Drop the existing future, catching any panics.
-        let result = panic::catch_unwind(AssertUnwindSafe(|| {
-            ptr::drop_in_place(self.boxed.as_ptr());
-        }));
-
-        // Overwrite the future behind the pointer. This is safe because the
-        // allocation was allocated with the same size and alignment as the type F.
-        let self_ptr: *mut F = self.boxed.as_ptr() as *mut F;
-        ptr::write(self_ptr, future);
-
-        // Update the vtable of self.boxed. The pointer is not null because we
-        // just got it from self.boxed, which is not null.
-        self.boxed = NonNull::new_unchecked(self_ptr);
-
-        // If the old future's destructor panicked, resume unwinding.
-        match result {
-            Ok(()) => {}
-            Err(payload) => {
-                panic::resume_unwind(payload);
-            }
-        }
+        let ptr = self.boxed.as_ptr();
+        let boxed = &mut self.boxed;
+
+        // Set new future and layout. By using DropFn we make sure the change
+        // will happen even if the existing future panics on drop.
+        let _set_layout = DropFn::new(move || {
+            // Overwrite the future behind the pointer. This is safe because the
+            // allocation was allocated with the same size and alignment as the type F.
+            let self_ptr: *mut F = boxed.as_ptr() as *mut F;
```

Comment:
I would prefer if this used the `ptr` variable rather than the mutable reference for aliasing reasons.